paderbox/array/sparse.py

```diff
@@ -0,0 +1,701 @@
+
+    def __copy__(self):
+        arr = self.__class__(shape=self.shape)
+        for segment in self._segments:
+            arr._add_segment(_SparseSegment(segment.onset, segment.array))
+        return arr
+
+    def __deepcopy__(self, memodict={}):
+        arr = self.__class__(shape=self.shape)
+        for segment in self._segments:
+            arr._add_segment(_SparseSegment(segment.offset, segment.array.copy()))
+        return arr
+
+    def __array__(self, dtype=None):
+        if self._is_torch:
+            raise NotImplementedError(
+                '__array__ is not implemented for torch SparseArrays'
+            )
+        return self.as_contiguous(dtype)
+
+    def __repr__(self):
+        import textwrap
+        if self._segments:
+            content = '\n'.join(map(str, self._segments))
+            content = textwrap.indent(content, '  ')
+            content = '\n' + content + ',\n  '
+        else:
+            content = ''
+        if self.pad_value != 0:
+            p = f'pad_value={self.pad_value}, '
+        else:
+            p = ''
+        return f'{self.__class__.__name__}({content}{p}shape={self.shape})'
+
+    def __setitem__(self, item, value):
+        """
+        >>> a = SparseArray(10)
+        >>> a[:3] = 1
+        >>> a[3:5] = np.arange(2) + 1
+        >>> a[6:8] = np.ones(2, dtype=int)
+        >>> np.asarray(a)
+        array([1, 1, 1, 1, 2, 0, 1, 1, 0, 0])
+
+        >>> a = SparseArray(10)
+        >>> a[:100] = 1
+        >>> a
+        SparseArray(
+          _SparseSegment(onset=0, length=10: [1 1 1 1 1 1 1 1 1 1]),
+          shape=(10,))
+
+        # Multi-dimensional
+        >>> a = SparseArray((2, 10))
+        >>> a[..., :5] = 1
+        >>> a[..., 7:] = np.arange(6).reshape((2, 3))
+        >>> np.asarray(a)
+        array([[1, 1, 1, 1, 1, 0, 0, 0, 1, 2],
+               [1, 1, 1, 1, 1, 0, 0, 3, 4, 5]])
+        >>> a[0, 5:] = 2
+        Traceback (most recent call last):
+          ...
+        NotImplementedError: Only ellipsis is allowed for first dimensions in setitem
+
+
+        # Test some exceptions
+        >>> a = SparseArray(shape=10)
+        >>> a[:5] = np.ones(3)
+        Traceback (most recent call last):
+         ...
+        ValueError: Could not assign array with shape (3,) to SparseArray slice with size 5
+        >>> a[:5] = np.ones((1, 5))
+        Traceback (most recent call last):
+         ...
+        ValueError: Shape mismatch: SparseArray has shape (10,), but assigned array has shape (1, 5).
+        >>> a[:100] = np.ones(100)
+        Traceback (most recent call last):
+         ...
+        ValueError: Could not assign array with shape (100,) to SparseArray slice with size 10
+        """
+        if isinstance(item, tuple):
+            selector = item[:len(self.shape) - 1]
+            if len(selector) != 1 or selector[0] is not ...:
+                raise NotImplementedError(
+                    'Only ellipsis is allowed for first dimensions in setitem'
```

Comment:
I don't get it. Do you want to limit `setitem` to be `[..., slice]` or `[slice]` (i.e. at most 2 arguments and if 2, the first must be Ellipsis?)?